devito/data/data.py

```diff
@@ -132,16 +140,76 @@ def _global(self, glb_idx, decomposition):
         ret._is_distributed = any(i is not None for i in decomposition)
         return ret
 
+    def check_slicing(func):
+        """Check if the operation will communication across MPI ranks."""
+        @wraps(func)
+        def wrapper(data, *args, **kwargs):
+            glb_idx = args[0]
+            if len(args) > 1 and isinstance(args[1], Data) \
+                    and args[1]._is_mpi_distributed:
+                mpi_slicing = True
+            elif data._is_mpi_distributed:
+                for i in as_tuple(glb_idx):
+                    if isinstance(i, slice) and i.step is not None and i.step < 0:
+                        mpi_slicing = True
+                        break
+                    else:
+                        mpi_slicing = False
+            else:
+                mpi_slicing = False
+            kwargs['mpi_slicing'] = mpi_slicing
+            return func(data, *args, **kwargs)
+        return wrapper
+
     @property
     def _is_mpi_distributed(self):
         return self._is_distributed and configuration['mpi']
 
     def __repr__(self):
         return super(Data, self._local).__repr__()
 
-    def __getitem__(self, glb_idx):
+    @check_slicing
+    def __getitem__(self, glb_idx, mpi_slicing=False):
         loc_idx = self._convert_index(glb_idx)
-        if loc_idx is NONLOCAL:
+        if mpi_slicing:
+            # Retrieve the pertinent local data prior to mpi send/receive operations
+            loc_data_idx = self._local_data_idx(loc_idx)
+            local_val = super(Data, self).__getitem__(loc_data_idx)
+
+            comm = self._distributor.comm
+            rank = self._distributor.myrank
+
+            owners, send, global_si, local_si = \
+                self._mpi_index_maps(loc_idx, local_val.shape, comm, rank)
+
+            retval = Data(local_val.shape, local_val.dtype.type,
+                          decomposition=local_val._decomposition,
+                          modulo=local_val._modulo,
+                          distributor=local_val._distributor)
+            it = np.nditer(owners, flags=['refs_ok', 'multi_index'])
+            while not it.finished:
+                index = it.multi_index
+                if rank == owners[index] and rank == send[index]:
```

Comment:
Add comment within each ``if`` ``elif`` ``else`` of this block explaining what is exactly going on?